### ISSUE TITLE
Root cause source of massive latency improvement in quantized MobileBertSquad benchmarks

### ISSUE THREAD

**Issue by julianwa** [google/iree#9170]
![image](https://user-images.githubusercontent.com/309732/169420189-58c35916-d31c-43d5-97cf-44cfa8d690e7.png)

We think it might be due to this commit: https://github.com/google/iree/commit/395d248e8bb9c7c221ec2e0dd18b833074e1f336

**Comment by mariecwhite** [comment]
The accuracy test for this model in iree-samples is passing, so these latency improvements look legitimate.

**Comment by bjacob** [comment]
This benchmark use to be spending over half of its time in Softmax, issue #8974, so the pigeonhole principle lets one guess that there may be an intersection here, so i'll prioritize looking into this before further work on softmax.

**Comment by julianwa** [comment]
heads up that @rsuderman thought that https://reviews.llvm.org/D125592 might have resulted in an improvement, but didn't think it could produce one this large

**Comment by benvanik** [comment]
Well that'd make a lot of sense! Nice! :P
We should do a quick scrub and see if any of the other ops with verbose expansions are in the models we are benchmarking. I see some others in the lit test like tosa.cast f32->i32 and it'd be useful to know if it ends up as llvm's fptosi, etc.

**Comment by hanhanW** [comment]
I looked into Tracy profiling. We got huge improvements for some dispatches. The execution time of dispatch_36 is cut from 466 ms to 30 ms.

The IR before submodule bump: https://gist.githubusercontent.com/hanhanW/a94cb5b6c9d151cedd379a97b6bc34b5/raw

The IR after submodule bump: https://gist.githubusercontent.com/hanhanW/6276e4c036055d3837d7359dbd425fbc/raw

There were scf ops within the Linalg op body, which stops vectorization. After submodule bumps, they are replaced with a `math.ctlz` op. In this context, my guess is that we're able to vectorize the dispatch. And it brings perf improvements to us.

I'm surprised that there are scf ops in a generic op. It would make vectorization harder. We should revisit if there are other cases like this.

**Comment by stellaraccident** [comment]
Are we at the point yet where we can have a warning issued if we fail to vectorize and use a fallback path?

**Comment by stellaraccident** [comment]
Thanks for the explanation. No need to change the order of operations.. I'm just looking forward to the near future where failure to vectorize can be considered warning worthy (maybe behind a flag that lets it be silenced). That will help us with regressions and odd cases without expensive diagnosis

**Comment by bjacob** [comment]
Thanks @hanhanW for the investigation. The IR here is the implementation of Softmax in TOSA, generated during TFLite-to-TOSA [here](https://github.com/tensorflow/tensorflow/blob/f8e703259b83f23ef2115edcc2bc2eff0ead1300/tensorflow/compiler/mlir/tosa/transforms/legalize_common.cc#L1313).

Latencies for this benchmark on Moto Edge X30:
* Before: 472 ms
* Now: 258 ms
* With dequantization of Softmax: 204 ms

So this is now much better than before -- but just dequantizing Softmax (ie making it use the float Softmax implementation) would still be another 20% speedup over this.

Here's how this currently looks in Tracy on Moto Edge X30:
![timeline-after](https://user-images.githubusercontent.com/79535/169582786-64989740-a278-44cd-9829-9c32b498668a.png)

* `main_dispatch_14` is unrelated
* `main_dispatch_59` is the computation of the `int32` exponential values. It is implemented as two table-lookups producing an `int16` value each, plus bit-operations combining those into the resulting `int32` value. The above Tracy timeline shows it accounting for the majority of the cost of Softmax (now that the subsequent arithmetic in `main_dispatch_61`, we get to that below, is less bad than it used to be). Zooming in on `main_dispatch_59` (MLIR TOSA source on the left, Aarch64 asm on the right): ![main_dispatch_59_after](https://user-images.githubusercontent.com/79535/169583225-157b1152-734d-4111-abc9-5439b4aa622c.png)

* `main_dispatch_60` is the computation of the sum of those exponentials (with a suitable right-shift applied), which becomes the denominator in the division implementing the softmax in `main_dispatch_61` below. It's only a small fraction of overall time according the the above timeline, so maybe not worth discussion. Here's view anyway:
![main_dispatch_60_after](https://user-images.githubusercontent.com/79535/169583624-a952f342-3c69-4c9c-8588-d343175feaf2.png)

* `main_dispatch_61` is the fixed-point division, where the numerator is the exponential value (from `main_dispatch_59`) and the denominator is the sum of exponentials (from `main_dispatch_60`).  It is what just became much faster. Before, this was un-vectorized code. Now this is vectorized, but in a naive way: the 32bit fixed-point multiplications are still naively done as multiplying into `int64` temporaries, so SIMD here is really only 2 lanes. It's still more than a 2x speedup due to how bad the non-SIMD code was (see disasm in #8974). But properly using fixed-point instructions would still be more than a 2x further speedup on top of that --- 2x increase in number of lanes, plus condensing multiple instructions (multiplication, right-shift...) into one fixed-point multiplication. That is the topic of Issue #9109.

**Comment by hanhanW** [comment]
>  I'm just looking forward to the near future where failure to vectorize can be considered warning worthy (maybe behind a flag that lets it be silenced)

good point, I'll add a flag and pass for warning it. It's helpful to see if there are heavy computations not getting vectorized.
